proposals/0391-replace-stake-program-floating-point.md

```diff
@@ -0,0 +1,138 @@
+   migration independently allows p-stake to enforce semantic
+   equivalence for its migration.
+
+## Requirements
+
+The new implementation must be a replacement that precisely models the
+intent of the original logic. Any resulting differences in output
+should be minor and a direct result of improved numerical precision.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+### Proposed Fixed-Point Implementation
+
+This proposal replaces `f64` with _rational arithmetic_, expressing
+the `warmup_cooldown_rate` as a fraction and reordering operations to
+preserve precision while using integer arithmetic.
+
+- The floating-point rates will be converted to their fractional
+  equivalents:
+  - `DEFAULT_WARMUP_COOLDOWN_RATE` (`0.25`) becomes a fraction of
+    **(numerator: 1, denominator: 4)**.
+  - `NEW_WARMUP_COOLDOWN_RATE` (`0.09`) becomes a fraction of
+    **(numerator: 9, denominator: 100)**.
+
+- The current flow
+  `(account_portion / cluster_portion) * (cluster_effective * rate)` is
+  reordered to
+  `(account_portion * cluster_effective * rate_numerator) / (cluster_portion * rate_denominator)`.
+  Instead of performing divisions early in the process (which truncates
+  intermediate results), all multiplications are performed first.
+
+- All intermediate multiplications are performed using `u128` to
+  prevent overflow.
+
+### State Compatibility
+
+To maintain backwards compatibility with on-chain stake account data,
+the `Delegation` struct will be modified:
+
+```diff
+pub struct Delegation {
+    pub voter_pubkey: Pubkey,
+    pub stake: u64,
+    pub activation_epoch: Epoch,
+    pub deactivation_epoch: Epoch,
+-   pub warmup_cooldown_rate: f64,
++    _reserved: [u8; 8],
+}
+```
+
+This preserves the exact memory size and layout of existing accounts.
+It is a legacy field anyway, with the actual rate being determined
+dynamically in functions.
+
+## Alternatives Considered
+
+Decimal scaling factor. Uses a uniform scaling factor represent rates
+as integers like BPS (e.g 0.25 = 2500). Rejected as rational
+arithmetic is more precise. It may be easier to use for external
+consumers, but these values are really only used internally to the
+stake interface crate.
+
+## Impact
+
+- **Stake Program:** The program must be updated to use the new
+  integer-based calculation helpers from `solana-stake-interface`. It is
+  doing so mostly through its use of
+  `stake.delegation.stake_activating_and_deactivating()`. Also, the new
+  `Delegation` struct definition (with its private field), will likely
+  impact how its being instantiated in a few areas.
+
+- **Agave:** Update the workspace dependency on
+  `solana-stake-interface`. Runtime stake processing already funnels
+  through Delegation::stake_activating_and_deactivating, so a dependency
+  bump automatically picks up the fixed-point math without touching
+  Agave code.
+
+- **Firedancer:** Will also need to update their stake interface
+  dependency in lockstep with Agave.
```

Comment:
Firedancer uses their own implementation of everything, so they'll need to implement the logic outlined in the previous section